honeycomb/honeycomb.go

```diff
@@ -0,0 +1,161 @@
+	"fmt"
+	"github.com/google/uuid"
+	"google.golang.org/grpc/codes"
+	"time"
+
+	libhoney "github.com/honeycombio/libhoney-go"
+	"go.opentelemetry.io/sdk/trace"
+)
+
+// Exporter is an implementation of trace.Exporter that uploads a span to Honeycomb
+type Exporter struct {
+	Builder        *libhoney.Builder
+	SampleFraction float64
+	// Service Name identifies your application. While optional, setting this
+	// field is extremely valuable when you instrument multiple services. If set
+	// it will be added to all events as `service_name`
+	ServiceName string
+}
+
+// Annotation represents an annotation with a value and a timestamp.
+type Annotation struct {
+	Timestamp time.Time `json:"timestamp"`
+	Value     string    `json:"value"`
+}
+
+// Span is the format of trace events that Honeycomb accepts
+type Span struct {
+	TraceID         string       `json:"trace.trace_id"`
+	Name            string       `json:"name"`
+	ID              uint64       `json:"trace.span_id"`
+	ParentID        uint64       `json:"trace.parent_id,omitempty"`
+	DurationMs      float64      `json:"duration_ms"`
+	Timestamp       time.Time    `json:"timestamp,omitempty"`
+	Annotations     []Annotation `json:"annotations,omitempty"`
+	Status          string       `json:"response.status_code,omitempty"`
+	Error           bool         `json:"error,omitempty"`
+	HasRemoteParent bool         `json:"has_remote_parent"`
+}
+
+// Close waits for all in-flight messages to be sent. You should
+// call Close() before app termination.
+func (e *Exporter) Close() {
+	libhoney.Close()
+}
+
+// NewExporter returns an implementation of trace.Exporter that uploads spans to Honeycomb
+//
+// writeKey is your Honeycomb writeKey (also known as your API key)
+// dataset is the name of your Honeycomb dataset to send trace events to
+//
+// Don't have a Honeycomb account? Sign up at https://ui.honeycomb.io/signup
+func NewExporter(apiKey, dataset string) *Exporter {
+	// Developer note: bump this with each release
+	versionStr := "1.0.1"
+	libhoney.UserAgentAddition = "Honeycomb-OpenTelemetry-exporter/" + versionStr
+
+	libhoney.Init(libhoney.Config{
+		WriteKey: apiKey,
+		Dataset:  dataset,
+	})
+	builder := libhoney.NewBuilder()
+	// default sample reate is 1: aka no sampling.
+	// set sampleRate on the exporter to be the sample rate given to the
+	// ProbabilitySampler if used.
+	return &Exporter{
+		Builder:        builder,
+		SampleFraction: 1,
+		ServiceName:    "",
+	}
+}
+
+// ExportSpan exports a SpanData to Jaeger.
+func (e *Exporter) ExportSpan(data *trace.SpanData) {
+	ev := e.Builder.NewEvent()
+	if e.SampleFraction != 0 {
+		ev.SampleRate = uint(1 / e.SampleFraction)
+	}
+	if e.ServiceName != "" {
+		ev.AddField("service_name", e.ServiceName)
+	}
+	// ev.Timestamp = sd.StartTime
+	hs := honeycombSpan(data)
+	ev.Add(hs)
+
+	// Add an event field for each attribute
+	// Hrm. Seems like attributes are trying to tell us something about the type of span
+	// ev.Add(data.Attributes)
```

Comment:
we definitely need the attributes to all be added...

Reply:
Agreed. I'd like to do that in a separate PR so that we can get the base exporter up and running first. I'll open [a ticket](https://app.asana.com/0/1136461855218693/1139131448927449?du=579988137396441)